File: rust/crates/orca-parity/src/modules/base_ref_search_result.rs
//! Parity dispatch for `orca_core::base_ref_search_result` vs
//! `src/shared/base-ref-search-result.ts`.

use orca_core::base_ref_search_result::{
    derive_legacy_local_branch_name, legacy_base_ref_search_result,
};
use serde_json::{json, Value};

pub fn dispatch(function: &str, input: &Value) -> Value {
    match function {
        "deriveLegacyLocalBranchName" => match input.as_str() {
            Some(ref_name) => Value::String(derive_legacy_local_branch_name(ref_name)),
            None => json!({ "__parity_error__": "expected string input" }),
        },
        "legacyBaseRefSearchResult" => match input.as_str() {
            Some(ref_name) => {
                // Match `JSON.stringify` of the TS `BaseRefSearchResult` object.
                let result = legacy_base_ref_search_result(ref_name);
                json!({
                    "refName": result.ref_name,
                    "localBranchName": result.local_branch_name,
                })
            }
            None => json!({ "__parity_error__": "expected string input" }),
        },
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}

File: src/renderer/src/lib/git-wasm/base-ref-search-result.ts
// Legacy base-ref search-result derivation, driven by the Rust orca-core
// base_ref_search_result port in the orca-git wasm module (the shared TS impl
// was deleted). Both consumers call this via `refs.map(...)` AFTER an RPC
// await, so wasm is ready; the not-ready branch still returns a correct minimal
// object (identity — no known remote prefix stripped) rather than null, so a
// `.map` over refs can never yield null entries during the wasm-boot window.
import { isGitWasmReady } from './git-line-stats'
import { orcaDispatch } from './orca_git_wasm.js'
import type { BaseRefSearchResult } from '../../../../shared/types'

function op(fn: string, input: unknown): unknown | null {
  if (!isGitWasmReady()) return null
  return JSON.parse(orcaDispatch('base-ref-search-result', fn, JSON.stringify(input ?? null)))
}

export function legacyBaseRefSearchResult(refName: string): BaseRefSearchResult {
  const r = op('legacyBaseRefSearchResult', refName) as BaseRefSearchResult | null
  return r ?? { refName, localBranchName: refName }
}
